BTProtocol/BTProtocol/BitTorrent/DownloadingTask.cs

```diff
@@ -0,0 +1,162 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading.Tasks;
+using System.Threading;
+
+using static BTProtocol.BitTorrent.MessageType;
+
+namespace BTProtocol.BitTorrent
+{
+    public class DownloadingTask : TorrentTask
+    {
+
+        TFData torrent_data;
+        Peer peer;
+        TcpClient client;
+
+        public DownloadingTask(TFData tfdata)
+        {
+            torrent_data = tfdata;
+            client = new TcpClient();
+        }
+
+        public void ExitThread()
+        {
+            MainProc.thread_pool.Release();
+        }
+
+        public void StartTask()
+        {
+            /*
+             * Downloading Task initialization steps:
+             *      - Check whether there remains to be any pieces to be downloaded
+             *      - Find an available, unvisted peer to connect to
+             *      - Connect to the peer and start downloading, or exit the Task if no peers are avaiable
+             */
+            // Call WaitOne to decrement the count of available threads.
+            MainProc.thread_pool.Wait();
+            // Release the lock on main so it can continue execution.
+            MainProc.main_semaphore.Release();
+
+            // Check the TFData to see if there are pieces that stll need to be downloaded. 
+            while (!torrent_data.CheckDownloadStatus())
+            {
+                (String, int) peer_addr;
+                client = new TcpClient();
+                while (!client.Connected)
+                {
+                    peer = FindAvailablePeer();
+                    if (peer != null)
+                        InitiateConnection();
+                    else
+                    {
+                        ExitThread();
+                        return;
+                    }
+                }
+
+                try
+                {
+                    SendHandshake(torrent_data, peer.netstream);
+                    ReceiveHandshake(torrent_data, peer.netstream);
+                    SendBitField(torrent_data, peer.netstream);
+                    ReceivePacket();
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine(e.Message);
+                }
+            }
+
+            ExitThread();
+        }
+
+        public Peer FindAvailablePeer()
```

Comment:
`private Peer FindAvailablePeer()` same for InitiateConnection and ReceivePacket